File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        curMax, curMin = 1, 1

        for num in nums:
            curMax, curMin = max(curMax * num, curMin * num, num), min(curMax * num, curMin * num, num)
            res = max(curMax, res)
        
        return res
        

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int[] m  = new int[nums.length];
        m[0] = nums[0];
        int q = nums[0];

        for (int i = 1; i <= nums.length - 1; i++) {
            if (m[i - 1] < 0) {
                m[i] = nums[i];
            } else {
                m[i] = nums[i] * m[i - 1];
            }
            q = Math.max(q, m[i]);
        }
        
        return q;
    }
}